cpu/asm/asm.go

```diff
@@ -0,0 +1,275 @@
+// This is helpful since Go will not automatically convert Signed Int8 to Unsigned Int8 from constants.
+func SImm16(x int16) Imm16 {
+	return signedImm[int16, Imm16](x)
+}
+
+func signedImm[T int8 | int16, R Imm8 | Imm16](x T) R {
+	if x < 0 {
+		return (R(x) - 1) | 0x80
+	}
+	return R(x)
+}
+
+// Ref is a generic type constraint for the reference a Pointer Operand can hold.
+// Any Pointer can be a reference to another Operand that is either a:
+// * Register8/16
+// * Immediate8/16
+type Ref interface {
+	Register8 | Register16 | Immediate8 | Immediate16
+
+	Operand
+}
+
+func (Register8) operand() {}
+
+// String implements fmt.Stringer
+func (r Register8) String() string {
+	switch {
+	case r >= A && r <= L:
+		return registerStrs[r]
+
+	default:
+		return "<invalid register>"
+	}
+}
+
+func (Register16) operand() {}
+
+// String implements fmt.Stringer
+func (r Register16) String() string {
+	switch {
+	case r >= AF && r <= PC:
+		return compoundStrs[r]
+
+	case r == SP:
+		return "SP"
+
+	case r >= spMin && r <= spMax:
+		offset := int(r - SP)
+		if offset > 0 {
+			return "SP + " + strconv.Itoa(offset)
+		} else {
+			return "SP - " + strconv.Itoa(int(r-SP))
+		}
+
+	default:
+		return "<invalid register>"
+	}
+}
+
+func (Immediate8) operand() {}
+
+// String implements fmt.Stringer
+func (i Immediate8) String() string {
+	return fmt.Sprintf("$%X", uint8(i))
+}
+
+func (Immediate16) operand() {}
+
+// String implements fmt.Stringer
+func (i Immediate16) String() string {
+	return fmt.Sprintf("$%X", uint16(i))
+}
+
+// Pointer represents a Pointer to a Reference Operand.
+// In some operations, The gameoby CPU can increment or decrement Pointer value, which can be
+// represented by the Delta field.
+type Pointer[R Ref] struct {
+	Ref   R
+	Delta Delta
+}
+
+// Delta represents whether a Register Pointer should be incremented or Decremented on use.
+// This is to represent instructions like LD [HL+], A
+type Delta int
+
+const (
+	Plus  Delta = 1
+	None  Delta = 0
+	Minus Delta = -1
+)
+
+// String implements fmt.Stringer
+func (d Delta) String() string {
+	switch d {
+	case Minus:
+		return "-"
+
+	case Plus:
+		return "+"
+
+	default:
+		return ""
+	}
+}
+
+func (Pointer[_]) operand() {}
+
+// String implements fmt.Stringer
+func (p Pointer[_]) String() string {
+	return "[" + p.Ref.String() + p.Delta.String() + "]"
+}
+
+// Ptr is a helper function to construct a generic Pointer Ref, with an Optional Delta provided in the form of
+// a variadic argument list.
+func Ptr[R Ref](r R, delta ...Delta) Pointer[R] {
+	var d Delta
+	if len(delta) > 0 {
+		d = delta[0]
+	}
+
+	return Pointer[R]{
+		Ref:   r,
+		Delta: d,
+	}
+}
+
+// 8 Bit Registers
+const (
+	A Register8 = iota
+	F
+	B
+	C
+	D
+	E
+	H
+	L
+)
+
+// 16 Bit Registers
+const (
+	AF Register16 = iota
+	BC
+	DE
+	HL
+	// below are 16 bit registers only
+	PC
+
+	// We use this special enumeration value to let us adding a signed 8 bit integer to SP as a right-hand opeand
```

Comment:
// We use this special enumeration value to let us adding a signed 8 bit integer to SP as a right-hand opeand
```suggestion
	// We use this special enumeration value to let us add a signed 8 bit integer to SP as a right-hand operand
```